narwhals/stable/v1/dataframe.py

```diff
@@ -0,0 +1,203 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+from typing import Any
+from typing import Literal
+from typing import cast
+from typing import overload
+
+from narwhals.dataframe import DataFrame as NwDataFrame
+from narwhals.dataframe import LazyFrame as NwLazyFrame
+from narwhals.dependencies import get_polars
+from narwhals.exceptions import InvalidIntoExprError
+from narwhals.typing import IntoDataFrameT
+from narwhals.typing import IntoFrameT
+from narwhals.utils import Implementation
+from narwhals.utils import inherit_doc
+
+if TYPE_CHECKING:
+    from types import ModuleType
+
+    from typing_extensions import Self
+    from typing_extensions import TypeVar
+
+    from narwhals.dataframe import MultiColSelector
+    from narwhals.dataframe import MultiIndexSelector
+    from narwhals.stable.v1.series import Series
+    from narwhals.typing import SingleColSelector
+    from narwhals.typing import SingleIndexSelector
+
+    FrameT = TypeVar("FrameT", "DataFrame[Any]", "LazyFrame[Any]")
+    DataFrameT = TypeVar("DataFrameT", bound="DataFrame[Any]")
+    LazyFrameT = TypeVar("LazyFrameT", bound="LazyFrame[Any]")
+    SeriesT = TypeVar("SeriesT", bound="Series[Any]")
```

Comment:
These are also copy-pasted in `functions.py`, `series.py` and `utils.py`

I guess we are not importing from `narwhals.stable.v1.typing` because of the circular import, right?
should we make a `narwhals.stable.v1._types` or similar for these?